packages/client/test/unit/MessageFactory.test.ts

```diff
@@ -0,0 +1,220 @@
+                        partitionKey: 'mockPartitionKey'
+                    },
+                    0
+                )
+            }).rejects.toThrow('Invalid combination of "partition" and "partitionKey"')
+        })
+
+        it('no partition key: uses same partition for all messages', async () => {
+            const messageFactory = createMessageFactory()
+            const msg1 = await messageFactory.createMessage(CONTENT, {
+                timestamp: TIMESTAMP
+            })
+            const msg2 = await messageFactory.createMessage(CONTENT, {
+                timestamp: TIMESTAMP
+            })
+            expect(msg1!.messageId.streamPartition).toBe(msg2!.messageId.streamPartition)
+        })
+
+        it('same partition key maps to same partition', async () => {
+            const messageFactory = createMessageFactory()
+            const partitionKey = `mock-partition-key-${random(Number.MAX_SAFE_INTEGER)}`
+            const msg1 = await messageFactory.createMessage(CONTENT, {
+                timestamp: TIMESTAMP,
+                partitionKey
+            })
+            const msg2 = await messageFactory.createMessage(CONTENT, {
+                timestamp: TIMESTAMP,
+                partitionKey
+            })
+            expect(msg1!.messageId.streamPartition).toBe(msg2!.messageId.streamPartition)
+        })
+
+        it('numeric partition key maps to the partition if in range', async () => {
+            const messageFactory = createMessageFactory()
+            const partitionKey = 10
+            const msg = await messageFactory.createMessage(CONTENT, {
+                timestamp: TIMESTAMP,
+                partitionKey
+            })
+            expect(msg!.messageId.streamPartition).toBe(partitionKey)
+        })
+
+        it('numeric partition key maps to partition range', async () => {
+            const messageFactory = createMessageFactory()
+            const partitionOffset = 20
+            const msg = await messageFactory.createMessage(CONTENT, {
+                timestamp: TIMESTAMP,
+                partitionKey: PARTITION_COUNT + partitionOffset
+            })
+            expect(msg!.messageId.streamPartition).toBe(partitionOffset)
+        })
+
+        it('selected random partition in range when partition count decreases', async () => {
+            let partitionCount: number = MAX_PARTITION_COUNT - 1
+            const messageFactory = createMessageFactory({
+                getPartitionCount: async () => partitionCount
+            })
+            while (partitionCount > 0) {
+                const msg = await messageFactory.createMessage(CONTENT, {
+                    timestamp: TIMESTAMP
+                })
+                expect(msg.messageId.streamPartition).toBeLessThan(partitionCount)
+                // eslint-disable-next-line no-plusplus
+                partitionCount--
+            }
+        })
+    })
+
+    describe('message chains', () => {
```

Comment:
Consider adding test case for the following publish two message with same msgChainId but different partitions. They should have the same `msgChainId` value but _not_ be linked (i.e. `prevMsgRef`  is null)